app/javascript/mastodon/features/interaction_modal/index.jsx

```diff
@@ -1,103 +1,261 @@
 
+const messages = defineMessages({
+  loginPrompt: { id: 'interaction_modal.login.prompt', defaultMessage: 'Domain of your home server, e.g. mastodon.social' },
+});
+
 const mapStateToProps = (state, { accountId }) => ({
   displayNameHtml: state.getIn(['accounts', accountId, 'display_name_html']),
-  signupUrl: state.getIn(['server', 'server', 'registrations', 'url'], null) || '/auth/sign_up',
 });
 
 const mapDispatchToProps = (dispatch) => ({
   onSignupClick() {
-    dispatch(closeModal({
-      modalType: undefined,
-      ignoreFocus: false,
-    }));
-    dispatch(openModal({ modalType: 'CLOSED_REGISTRATIONS' }));
+    dispatch(closeModal());
+    dispatch(openModal('CLOSED_REGISTRATIONS'));
   },
 });
 
-class Copypaste extends PureComponent {
+const PERSISTENCE_KEY = 'mastodon_home';
+
+const isValidDomain = value => {
+  const url = new URL('https:///path');
+  url.hostname = value;
+  return url.hostname === value;
+};
+
+const valueToDomain = value => {
+  // If the user uses a URL to their profile page or server
+  if (/^https?:\/\//.test(value)) {
+    try {
+      return (new URL(value)).host;
+    } catch {
+      return '';
+    }
+  // If the user writes their full handle including username
+  } else if (value.includes('@')) {
+    const segments = value.split('@');
+
+    if (segments.length === 2) {
+      return segments[1];
+    } else {
+      return '';
+    }
+  }
+
+  return value;
+};
+
+class LoginForm extends React.PureComponent {
 
   static propTypes = {
-    value: PropTypes.string,
+    resourceUrl: PropTypes.string,
+    intl: PropTypes.object.isRequired,
   };
 
   state = {
-    copied: false,
+    value: localStorage ? (localStorage.getItem(PERSISTENCE_KEY) || '') : '',
+    expanded: false,
+    selectedOption: -1,
+    isLoading: false,
+    isSubmitting: false,
+    error: false,
+    options: [],
   };
 
   setRef = c => {
     this.input = c;
   };
 
-  handleInputClick = () => {
-    this.setState({ copied: false });
-    this.input.focus();
-    this.input.select();
-    this.input.setSelectionRange(0, this.input.value.length);
+  handleChange = ({ target }) => {
+    this.setState({ value: target.value, isLoading: true, error: false }, () => this._loadOptions());
+  };
+
+  handleSubmit = () => {
+    const { value } = this.state;
+    const { resourceUrl } = this.props;
+
+    const domain = valueToDomain(value);
+
+    if (!isValidDomain(domain)) {
+      this.setState({ error: true });
+      return;
+    }
+
+    if (localStorage) {
+      localStorage.setItem(PERSISTENCE_KEY, domain);
+    }
+
+    const fallbackUrl = `https://${domain}/authorize_interaction?uri={uri}`;
+
+    this.setState({ isSubmitting: true });
+
+    api().get(`https://${domain}/.well-known/webfinger`, { params: { resource: domain } }).then(({ data }) => {
+      const template = data.links.find(link => link.rel === 'http://ostatus.org/schema/1.0/subscribe')?.template;
```

Comment:
does this bit only pick up on the ostatus rel and the `template` value? i can't find a definition for `template`, and it's not in RFC 7033, so i'd like to know what you think about the proposal for extending as:proxyUrl

Reply:
It feels weird to use the path to a JSON-LD property as a `rel` value. Maybe we could use something from [here](https://microformats.org/wiki/existing-rel-values), if we're adding new properties anyway? There seems to even be something called `respond-proxy`, which sounds similar to our use case.

Reply:
the iri for proxyUrl is used to define semantics, and these semantics 90% match what is going on here -- you just need to handle GET in addition to POST (or in mastodon's terms, add support for POST in addition to GET, and the `id` parameter instead of `uri`). it's not weird to use a semantic iri as a rel-value, and arguably if you consider RDF then using it as a property creates a qualified relation anyway, so it's the same thing.